adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+		}
+		duration = getDuration(int64(content.MetaData.VideoInfo.VideoDuration))
+	}
+
+	var adTitle, _ = getDecodeValue(content.MetaData.Title)
+	var adId = content.Contentid
+	var creativeId = content.Contentid
+	var trackingEvents strings.Builder
+	var dspImpTracking2Str = ""
+	var dspClickTracking2Str = ""
+	var errorTracking2Str = ""
+	if content.Monitor != nil {
+		for _, monitor := range content.Monitor {
+			if len(monitor.Url) == 0 {
+				continue
+			}
+			var event = ""
+			switch monitor.EventType {
+			case "vastError":
+				errorTracking2Str = getVastImpClickErrorTrackingUrls(monitor.Url, "vastError")
+			case "imp":
+				dspImpTracking2Str = getVastImpClickErrorTrackingUrls(monitor.Url, "imp")
+			case "click":
+				dspClickTracking2Str = getVastImpClickErrorTrackingUrls(monitor.Url, "click")
+			case "userclose":
+				event = "skip&closeLinear"
+			case "playStart":
+				event = "start"
+			case "playEnd":
+				event = "complete"
+			case "playResume":
+				event = "resume"
+			case "playPause":
+				event = "pause"
+			case "soundClickOff":
+				event = "mute"
+			case "soundClickOn":
+				event = "unmute"
+			default:
+			}
+			if event != "" {
+				if event != "skip&closeLinear" {
+					trackingEvents.WriteString(getVastEventTrackingUrls(monitor.Url, event))
+				} else {
+					trackingEvents.WriteString(getVastEventTrackingUrls(monitor.Url, "skip&closeLinear"))
+				}
+			}
+		}
+	}
+
+	adm = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
+		"<VAST version=\"3.0\">" +
+		"<Ad id=\"" + adId + "\"><InLine>" +
+		"<AdSystem>HuaweiAds</AdSystem>" +
+		"<AdTitle>" + adTitle + "</AdTitle>" +
+		errorTracking2Str + dspImpTracking2Str +
+		"<Creatives>" +
+		"<Creative adId=\"" + adId + "\" id=\"" + creativeId + "\">" +
+		"<Linear>" +
+		"<Duration>" + duration + "</Duration>" +
+		"<TrackingEvents>" + trackingEvents.String() + "</TrackingEvents>" +
+		"<VideoClicks>" +
+		"<ClickThrough><![CDATA[" + clickUrl + "]]></ClickThrough>" +
+		dspClickTracking2Str +
+		"</VideoClicks>" +
+		"<MediaFiles>" +
+		"<MediaFile delivery=\"progressive\" type=\"" + mime + "\" width=\"" + strconv.Itoa(int(adWidth)) + "\" " +
+		"height=\"" + strconv.Itoa(int(adHeight)) + "\" scalable=\"true\" maintainAspectRatio=\"true\"> " +
+		"<![CDATA[" + resourceUrl + "]]>" +
+		"</MediaFile>" +
+		"</MediaFiles>" +
+		"</Linear>" +
+		"</Creative>" +
+		"</Creatives>" +
+		"</InLine>" +
+		"</Ad>" +
+		"</VAST>"
+	return adm, adWidth, adHeight, nil
+}
+
+func getVastImpClickErrorTrackingUrls(urls []string, eventType string) (result string) {
+	var trackingUrls strings.Builder
+	for _, url := range urls {
+		if eventType == "click" {
+			trackingUrls.WriteString("<ClickTracking><![CDATA[" + url + "]]></ClickTracking>")
+		} else if eventType == "imp" {
+			trackingUrls.WriteString("<Impression><![CDATA[" + url + "]]></Impression>")
+		} else if eventType == "vastError" {
+			trackingUrls.WriteString("<Error><![CDATA[" + url + "&et=[ERRORCODE]]]></Error>")
+		}
```

Comment:
To make the most use of the string builder, please break up the string concatenation into separate WriteString calls. Same for getVastEventTrackingUrls.

Reply:
corrected